hpcflow/sdk/core/actions.py

```diff
@@ -2377,21 +2458,21 @@ def get_precise_scope(self) -> app.ActionScope:
             )
 
         if self.input_file_generators:
-            return self.app.ActionScope.input_file_generator(
+            return self._app.ActionScope.input_file_generator(
                 file=self.input_file_generators[0].input_file.label
             )
         elif self.output_file_parsers:
             if self.output_file_parsers[0].output is not None:
-                return self.app.ActionScope.output_file_parser(
+                return self._app.ActionScope.output_file_parser(
                     output=self.output_file_parsers[0].output
                 )
             else:
-                return self.app.ActionScope.output_file_parser()
+                return self._app.ActionScope.output_file_parser()
         else:
-            return self.app.ActionScope.main()
+            return self._app.ActionScope.main()
 
     def is_input_type_required(
-        self, typ: str, provided_files: List[app.FileSpec]
+        self, typ: str, provided_files: Container[FileSpec]
```

Comment:
I think using a set instead of a list for `provided_files` means the membership test `ifg.input_file not in provided_files` below fails with `TypeError: unhashable type: 'FileSpec'`. I have just committed two two failing tests (8fb13758d44826f2bc595bada746b57c95a19dd2) that test for the correct behaviour.

Reply:
Which turned out to be _slightly_ tricky to do. Fortunately, it seems that `FileNameSpec` is treated as a read-only class so it doesn't need a `list` inside (a `tuple` does fine). I couldn't find any examples of `FileNameSpec.args` being non-empty, but I'll assume it's possible.